### PR TITLE
Disconnect device on Ctrl+C in `nbd lo`

### PR DESCRIPTION
**Pull request by bduffany**
Fixes https://github.com/Merovius/nbd/issues/13

I verified that the disconnect is successful, as I am able to re-run the script without any errors (previously, re-running would fail with "device or resource busy" since presumably the old device was still connected).

Here are the kernel logs showing the successful disconnect:

```
[85677.085648] block nbd2: NBD_DISCONNECT
[85677.085675] block nbd2: Disconnected due to user request.
[85677.085681] block nbd2: shutting down sockets
```

However, there is a separate bug (to be fixed in a future PR?) where we fail to validate the netlink response:

`2023/06/01 10:26:54 Error while disconnecting /dev/nbd2: netlink validate: mismatched sequence in netlink reply`

The validation failure does not seem to matter since the device is already disconnected, but it's probably worth fixing.

### CODE REVIEW

**Comment by Merovius** [comment]
I'm not sure this is the best way to solve this. `Loopback` is currently documented to return when its `context.Context` is cancelled. So I think a better way to do this would be to use `signal.NotifyContext` to cancel the context. That way `(*loCmd).Execute` would just wait for `wait()` to return as normal, instead of doing its own signalling.

I don't think `Loopback` currently disconnects the device, though, if the context is cancelled. But it should. This may require a bit of juggling to make sure there is still a non-cancelled context for the cleanup (though it doesn't really seem to respect cancellation anyways right now m( ). Also, `Lookup` doesn't seem to clean up the socket pair it opens.

In other words, I think this might require a bit more cleanup and rejigging. Do you feel up to that? Otherwise I will do it myself.

**Comment by bduffany** [comment]
> I'm not sure this is the best way to solve this. Loopback is currently documented to return when its context.Context is cancelled. So I think a better way to do this would be to use signal.NotifyContext to cancel the context. That way (*loCmd).Execute would just wait for wait() to return as normal, instead of doing its own signalling.

I agree - TIL about `signal.NotifyContext`. Done :+1: 

> I don't think Loopback currently disconnects the device, though, if the context is cancelled. But it should.

This seems reasonable to me. Done :+1:

> Also, Lookup doesn't seem to clean up the socket pair it opens.

It looks like both `server.Close()` and `client.Close()` get called assuming the user properly calls `.wait()`? Is there some other cleanup that needs to happen? (Sorry, I am relatively new to `socketpair()`, and couldn't find much after some googling)

File: netlink.go [hunk 1]
```diff
@@ -67,26 +68,32 @@ func Loopback(ctx context.Context, d Device, size uint64) (idx uint32, wait func
...
-		client.Close()
-	}()
 	go func() {
 		err := serve(ctx, serverc, connParameters{exp, defaultBlockSizes})
 		if e := ctx.Err(); e != nil {
 			err = e
 		}
 		cancel()
-		ch <- err
+		client.Close()
```
Comment (bduffany) [inline netlink.go:82]: I moved the `client.Close()` here to ensure that `wait()` waits for it to happen. Otherwise I think this could've potentially been called after the process closed, though it doesn't really matter since the OS would've freed it anyway. Either way, it seemed simpler to just have one goroutine that does all the cleanup?

File: netlink.go [hunk 2]
```diff
@@ -67,26 +68,32 @@ func Loopback(ctx context.Context, d Device, size uint64) (idx uint32, wait func
...
 		serverc.Close()
+		if <-configErr == nil {
```
Comment (bduffany) [inline netlink.go:84]: This code felt slightly unconventional to me, but the other idea I had of having a `configured bool` that gets set to true (after `Configure()` is called successfully) seemed like it could have a race condition. Definitely open to other ideas here.

**Comment by Merovius** [comment]
I'll take a look in a bit. FWIW I might've been wrong about closing the socket pair, I just couldn't see it at my casual glance this morning

**Comment by Merovius** [comment]
Okay, I like how the changes to `cmd/ndb` look now :) I find how `Loopback` is written very confusing though. That's mostly on me, the existing code is a bit of a mess.

I think the code can be made significantly easier to read by 1. moving the `Configure` call up a bit (so we don't need the `configErr` channel) and 2. using an [errgroup](https://pkg.go.dev/golang.org/x/sync/errgroup) to do the bookkeeping of the goroutine for us. `wait` can then, in sequence, 1. `Wait()` on the `errgroup`, 2. `Disconnect`, 3. close the client/server sockets. And return the first error that happens.

Does that seem right to you?

**Comment by bduffany** [comment]
Yep, that seems right. One thing I found is that if I return the first error that happens, the error returned will always be `ContextCanceled`. So to make sure we can still return a useful error, I just accumulated all the errors in a slice and returned them using `errors.Join()`. I figured this way, people can check the errors if they want, and if the only error is ContextCanceled (and they intentionally canceled the context), then they can ignore the error. Does that seem reasonable?

**Comment by Merovius** [comment]
Not a fan of `errors.Join` - its multi-line output format doesn't play well with common logging patterns. I instead decided to just ignore cancellation errors, as they are not actionable. And move up Disconnect, as an error from that is probably most important.

Thanks for the PR and your patience :+1:
